File: languages/java/runtime/src/main/java/co/revaly/sdk/testing/SyntheticData.java
package co.revaly.sdk.testing;

/**
 * Synthetic response bodies for the mock transport — SYNTHETIC DATA ONLY (ADR-SDK-020): no real
 * PANs, customers, or keys ever appear here, and no card fields at all (the mock exercises the
 * failover contract, which never needs them).
 */
public final class SyntheticData {

    public static final String DEFAULT_MERCHANT_TRANSACTION_ID = "mtx-synthetic-1";
    public static final String DEFAULT_TRANSACTION_ID = "txn-synthetic-1";

    private SyntheticData() {}

    /** A terminal transaction record; {@code transactionStatus} 1=Approved 2=Declined 3=Error. */
    public static String transaction(int transactionStatus, String merchantTransactionId) {
        return "{"
                + "\"transactionId\":\""
                + escape(DEFAULT_TRANSACTION_ID)
                + "\","
                + "\"merchantTransactionId\":\""
                + escape(merchantTransactionId)
                + "\","
                + "\"transactionStatus\":"
                + transactionStatus
                + ","
                + "\"transactionType\":\"Charge\","
                + "\"responseCode\":\"00\","
                + "\"message\":\"synthetic\""
                + "}";
    }

    /** A pending payment intent (post-P-2 shape; {@code state} is the discriminator). */
    public static String pending(String merchantTransactionId) {
        return "{"
                + "\"state\":\"pending\","
                + "\"merchantTransactionId\":\""
                + escape(merchantTransactionId)
                + "\","
                + "\"transactionType\":\"Charge\""
                + "}";
    }

    /** A grouped envelope (the {@code includeAllTransactions=true} shape). */
    public static String transactionGroup(String merchantTransactionId) {
        String item = transaction(1, merchantTransactionId);
        return "{\"transaction\":" + item + ",\"transactions\":[" + item + "]}";
    }

    /** An {@code ErrorResponse} body; pass a null {@code code} to omit the field. */
    public static String errorBody(String error, String code) {
        if (code == null) {
            return "{\"error\":\"" + escape(error) + "\"}";
        }
        return "{\"error\":\"" + escape(error) + "\",\"code\":\"" + escape(code) + "\"}";
    }

    private static String escape(String value) {
        return value == null ? "" : value.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}

File: languages/php/runtime/src/Testing/SyntheticData.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Testing;

/**
 * Synthetic wire bodies for the mock transport. SYNTHETIC DATA ONLY (ADR-SDK-020):
 * every identifier is an obvious placeholder and no real PAN/CVV/PII ever appears here.
 */
final class SyntheticData
{
    public const DEFAULT_MERCHANT_TRANSACTION_ID = 'mtx-synthetic-1';
    public const DEFAULT_TRANSACTION_ID = 'txn-synthetic-1';
    public const DEFAULT_CORRELATION_ID = 'corr-synthetic-1';

    private function __construct()
    {
    }

    /** A terminal transaction record (`transactionStatus` 1=approved, 2=declined, 3=error). */
    public static function transaction(
        int $transactionStatus,
        string $merchantTransactionId = self::DEFAULT_MERCHANT_TRANSACTION_ID,
    ): string {
        return json_encode([
            'transactionId' => self::DEFAULT_TRANSACTION_ID,
            'transactionDate' => '2026-01-01T00:00:00Z',
            'transactionStatus' => $transactionStatus,
            'message' => 'synthetic outcome',
            'responseCode' => '00',
            'transactionType' => 'charge',
            'merchantTransactionId' => $merchantTransactionId,
            'currency' => 'USD',
            'amount' => 1999,
        ], JSON_THROW_ON_ERROR);
    }

    /** A pending intent record (post-P-2 shape; `state` is the discriminator). */
    public static function pending(
        string $merchantTransactionId = self::DEFAULT_MERCHANT_TRANSACTION_ID,
    ): string {
        return json_encode([
            'state' => 'pending',
            'merchantTransactionId' => $merchantTransactionId,
            'receivedAt' => '2026-01-01T00:00:00Z',
        ], JSON_THROW_ON_ERROR);
    }

    /** A grouped envelope (`include_all_transactions` shape). */
    public static function transactionGroup(
        string $merchantTransactionId = self::DEFAULT_MERCHANT_TRANSACTION_ID,
    ): string {
        return json_encode([
            'transactions' => [
                json_decode(self::transaction(1, $merchantTransactionId), true),
            ],
        ], JSON_THROW_ON_ERROR);
    }

    /** An ErrorResponse body; `code` is omitted when null (matching the wire contract). */
    public static function errorBody(string $error, ?string $code = null): string
    {
        $body = ['error' => $error];
        if ($code !== null) {
            $body['code'] = $code;
        }

        return json_encode($body, JSON_THROW_ON_ERROR);
    }
}
